src/file-events/cpp/win_fsnotifier.cpp

```diff
@@ -182,25 +183,18 @@ void Server::startWatching(JNIEnv* env, const u16string& path) {
         return;
     }
 
-    WatchPoint* watchPoint = new WatchPoint(this, path, directoryHandle);
-
     HANDLE threadHandle = watcherThread.native_handle();
-    int ret = watchPoint->awaitListeningStarted(threadHandle);
-    switch (ret) {
-        case WATCH_LISTENING:
-            watchPoints.push_back(watchPoint);
-            break;
-        default:
-            log_severe(env, "Couldn't start listening to '%ls': %d", pathW.c_str(), ret);
-            delete watchPoint;
-            // TODO Error handling
-            break;
-    }
+    watchPoints.emplace(piecewise_construct,
+        forward_as_tuple(path),
+        forward_as_tuple(this, path, directoryHandle, threadHandle));
 }
 
-void Server::reportFinished(WatchPoint* watchPoint) {
-    watchPoints.remove(watchPoint);
-    delete watchPoint;
+void Server::stopWatching(JNIEnv* env, const u16string& path) {
+    watchPoints.find(path)->second.close();
```

Comment:
Should we check here if we are watching the path?

Reply:
Yes, fixing.